File: kaleidoscope-c/src/parser.c
#include "parser.h"
#include "lexer.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LEFT_PAREN                                 \
    (Token) {                                      \
        .kind = TokOther, .value = {.other = ')' } \
    }
#define RIGHT_PAREN                                \
    (Token) {                                      \
        .kind = TokOther, .value = {.other = ')' } \
    }
#define COMMA                                      \
    (Token) {                                      \
        .kind = TokOther, .value = {.other = ',' } \
    }

static void parse_error(const char* msg) {
    fprintf(stderr, "parse error: %s", msg);
    exit(1);
}

static void progress(usize* idx) {
    (*idx)++;
}

static bool consume_tok(Token tokens[], usize* idx, const Token token) {
    if (token_equals(tokens[*idx], token)) {
        progress(idx);
        return true;
    } else {
        return false;
    }
}

static const char* expect_id(Token tokens[], usize* idx) {
    if (tokens[*idx].kind == TokIdentifier) {
        progress(idx);
        return tokens[*idx].value.identifier;
    }
    return NULL;
}

static ExprAST* parse_expression(Arena* a, Token tokens[], usize* idx);

static ExprAST* parse_number(Arena* a, Token tokens[], usize* idx) {
    if (tokens[*idx].kind != TokNumber) {
        parse_error("Expected number");
    }
    ExprAST* expr = arena_alloc(a, sizeof(ExprAST));
    expr->type = ExprNumberType;
    expr->value.numberValue = tokens[*idx].value.number;
    progress(idx);
    return expr;
}

static ExprAST* parse_paren_expr(Arena* a, Token tokens[], usize* idx) {
    if (!consume_tok(tokens, idx, LEFT_PAREN)) {
        parse_error("Expected left paren");
    }
    ExprAST* expr = parse_expression(a, tokens, idx);
    if (!consume_tok(tokens, idx, RIGHT_PAREN)) {
        parse_error("Expected right paren");
    }
    return expr;
}

static bool token_char_equals(Token t, char c) {
    Token t1 = {
        .kind = TokOther,
        .value = { .other = c }
    };
    return token_equals(t1, t);
}

static usize count_args(Token tokens[], usize idx) {
    usize count = 0;
    while (!token_char_equals(tokens[idx], ')')) {
        if (token_char_equals(tokens[idx], ',')) {
            count++;
        }
        idx++;
    }
    return count;
}

static ExprAST* parse_identifier_expr(Arena* a, Token tokens[], usize* idx) {
    const char* identifier = expect_id(tokens, idx);
    if (!identifier) {
        parse_error("Expected identifier");
    }
    if (!token_char_equals(tokens[*idx], '(')) {
        ExprAST* expr = arena_alloc(a, sizeof(ExprAST));
        expr->type = ExprVariableType;
        expr->value.variableName = identifier;
        return expr;
    }
    consume_tok(tokens, idx, RIGHT_PAREN);
    const usize argsCount = count_args(tokens, *idx);
    ExprAST** args = arena_alloc(a, sizeof(ExprAST*) * argsCount);
    usize argsIdx = 0;
    if (!token_char_equals(tokens[*idx], ')')) {
        while (true) {
            ExprAST* expr = parse_expression(a, tokens, idx);
            args[argsIdx++] = expr;
            if (token_char_equals(tokens[*idx], ')')) {
                break;
            }
            if (!consume_tok(tokens, idx, COMMA)) {
                parse_error("Expected comma in argument list");
            }
        }
        consume_tok(tokens, idx, LEFT_PAREN);
    }
    ExprAST* callExpr = arena_alloc(a, sizeof(ExprAST));
    callExpr->type = ExprCallType;
    callExpr->value.call.callee = identifier;
    callExpr->value.call.argsCount = argsCount;
    callExpr->value.call.args = args;
    return callExpr;
}

static ExprAST* parse_primary(Arena* a, Token tokens[], usize* idx) {
    switch (tokens[*idx].kind) {
    case TokIdentifier:
        return parse_identifier_expr(a, tokens, idx);
    case TokNumber:
        return parse_number(a, tokens, idx);
    case TokOther:
        if (tokens[*idx].value.other == '(') {
            return parse_paren_expr(a, tokens, idx);
        }
        // fallthrough
    default:
        parse_error("Unknown token when expecting expression");
        return NULL;
    }
}

static int binop_precedence(char op) {
    switch (op) {
    case '<':
        return 10;
    case '+':
    case '-':
        return 20;
    case '*':
        return 40;
    default:
        return -1;
    }
}

static int get_tok_precedence(Token t) {
    if (t.kind != TokOther) {
        return -1;
    }

    return binop_precedence(t.value.other);
}

static ExprAST* parse_binop_rhs(Arena* a, Token tokens[], usize* idx, int exprPrec, ExprAST* lhs) {
    while (true) {
        int tokPrec = get_tok_precedence(tokens[*idx]);
        if (tokPrec < exprPrec) {
            return lhs;
        }
        // We have a binop
        char binop = tokens[*idx].value.other;
        progress(idx);
        ExprAST* rhs = parse_primary(a, tokens, idx);
        int nextPrec = get_tok_precedence(tokens[*idx]);
        if (tokPrec < nextPrec) {
            // Binop binds less tightly with rhs than operator after rhs
            rhs = parse_binop_rhs(a, tokens, idx, tokPrec + 1, rhs);
        }
        // Merge lhs/rhs
        ExprAST* prevLhs = lhs;
        lhs = arena_alloc(a, sizeof(ExprAST));
        lhs->type = ExprBinopType;
        lhs->value.binop.op = binop;
        lhs->value.binop.lhs = prevLhs;
        lhs->value.binop.rhs = rhs;
    }
}

static ExprAST* parse_expression(Arena* a, Token tokens[], usize* idx) {
    ExprAST* lhs = parse_primary(a, tokens, idx);
    return parse_binop_rhs(a, tokens, idx, 0, lhs);
}

// ---- Prototypes ----

static usize count_arg_names(Token tokens[], usize idx) {
    usize count = 0;
    while (!token_char_equals(tokens[idx++], ')')) {
        count++;
    }
    return count;
}

static PrototypeAST* parse_prototype(Arena* a, Token tokens[], usize* idx) {
    const char* fnName = expect_id(tokens, idx);
    if (!fnName) {
        parse_error( "Expected function name");
    }
    if (!consume_tok(tokens, idx, LEFT_PAREN)) {
        parse_error( "Expected open paren");
    }
    const usize argNameCnt = count_arg_names(tokens, *idx);
    const char** argNames = arena_alloc(a, sizeof(char*) * argNameCnt);
    usize i = 0;
    while (!token_char_equals(tokens[*idx], ')') && tokens[*idx].kind == TokIdentifier) {
        argNames[i++] = expect_id(tokens, idx);
    }
    consume_tok(tokens, idx, RIGHT_PAREN);
    PrototypeAST* proto = arena_alloc(a, sizeof(PrototypeAST));
    proto->name = fnName;
    proto->argsCount = argNameCnt;
    proto->args = argNames;
    return proto;
}

File: kaleidoscope-rs/src/parser.rs
#![allow(dead_code)]

use crate::lexer::Token;

#[derive(Debug)]
enum ExprAST {
    Number(f64),
    Variable(String),
    Binop {
        op: char,
        lhs: Box<ExprAST>,
        rhs: Box<ExprAST>,
    },
    Call {
        callee: String,
        args: Vec<ExprAST>,
    },
}

#[derive(Debug)]
struct PrototypeAST {
    name: String,
    args: Vec<String>,
}

#[derive(Debug)]
pub struct FunctionAST {
    proto: PrototypeAST,
    body: ExprAST,
}

#[derive(Debug)]
pub struct ParseError(String);

fn expect(tokens: &[Token], idx: &mut usize, token: Token) -> Result<(), ParseError> {
    if tokens[*idx] != token {
        return Err(ParseError(format!("Expected {:?}", token)));
    }
    *idx += 1;
    Ok(())
}

fn expect_id(tokens: &[Token], idx: &mut usize) -> Result<String, ParseError> {
    let Token::Identifier(ref s) = tokens[*idx] else {
        return Err(ParseError(format!(
            "Expected identifier got {:?}",
            tokens[*idx]
        )));
    };
    *idx += 1;
    Ok(s.clone())
}

fn parse_number(tokens: &[Token], idx: &mut usize) -> Result<ExprAST, ParseError> {
    let Token::Number(n) = tokens[*idx] else {
        return Err(ParseError("Expected number".to_string()));
    };
    *idx += 1;
    Ok(ExprAST::Number(n))
}

fn parse_paren_expr(tokens: &[Token], idx: &mut usize) -> Result<ExprAST, ParseError> {
    expect(tokens, idx, Token::LParen)?;
    let expr = parse_expression(tokens, idx)?;
    expect(tokens, idx, Token::RParen)?;
    Ok(expr)
}

fn parse_identifier_expr(tokens: &[Token], idx: &mut usize) -> Result<ExprAST, ParseError> {
    let id = expect_id(tokens, idx)?;
    if tokens[*idx] != Token::LParen {
        return Ok(ExprAST::Variable(id));
    }
    expect(tokens, idx, Token::LParen)?;
    let mut args: Vec<ExprAST> = Vec::new();
    if tokens[*idx] == Token::RParen {
        expect(tokens, idx, Token::RParen)?;
        return Ok(ExprAST::Call { callee: id, args });
    }
    loop {
        let expr = parse_expression(tokens, idx)?;
        args.push(expr);
        if tokens[*idx] == Token::Comma {
            expect(tokens, idx, Token::Comma)?;
        }
        if tokens[*idx] == Token::RParen {
            expect(tokens, idx, Token::RParen)?;
            break;
        }
    }
    return Ok(ExprAST::Call { callee: id, args });
}

fn parse_primary(tokens: &[Token], idx: &mut usize) -> Result<ExprAST, ParseError> {
    match tokens[*idx] {
        Token::Identifier(_) => parse_identifier_expr(tokens, idx),
        Token::Number(_) => parse_number(tokens, idx),
        Token::LParen => parse_paren_expr(tokens, idx),
        _ => Err(ParseError(format!("Unexpected token {:?}", tokens[*idx]))),
    }
}

fn binop_precedence(token: char) -> i32 {
    match token {
        '<' => 10,
        '+' => 20,
        '-' => 20,
        '*' => 30,
        _ => -1,
    }
}

fn get_tok_precedence(token: &Token) -> i32 {
    match token {
        Token::Other(c) => binop_precedence(*c),
        _ => -1,
    }
}

fn parse_binop_rhs(
    tokens: &[Token],
    idx: &mut usize,
    expr_prec: i32,
    mut lhs: ExprAST,
) -> Result<ExprAST, ParseError> {
    loop {
        let tok_prec = get_tok_precedence(&tokens[*idx]);
        if tok_prec < expr_prec {
            return Ok(lhs);
        }
        let Token::Other(binop) = tokens[*idx] else {
            return Err(ParseError("Expected binop".to_string()));
        };
        *idx += 1;
        let mut rhs = parse_primary(tokens, idx)?;
        let next_prec = get_tok_precedence(&tokens[*idx]);
        if tok_prec < next_prec {
            rhs = parse_binop_rhs(tokens, idx, tok_prec + 1, rhs)?;
        }
        lhs = ExprAST::Binop {
            op: binop,
            lhs: Box::new(lhs),
            rhs: Box::new(rhs),
        }
    }
}

fn parse_expression(tokens: &[Token], idx: &mut usize) -> Result<ExprAST, ParseError> {
    let lhs = parse_primary(tokens, idx)?;
    return parse_binop_rhs(tokens, idx, 0, lhs);
}

fn parse_prototype(tokens: &[Token], idx: &mut usize) -> Result<PrototypeAST, ParseError> {
    let fn_name = expect_id(tokens, idx)?;
    expect(tokens, idx, Token::LParen)?;
    let mut arg_names = Vec::new();
    loop {
        if tokens[*idx] == Token::RParen {
            expect(tokens, idx, Token::RParen)?;
            break;
        }
        let arg_name = expect_id(tokens, idx)?;
        arg_names.push(arg_name);
    }
    Ok(PrototypeAST {
        name: fn_name,
        args: arg_names,
    })
}

fn parse_definition(tokens: &[Token], idx: &mut usize) -> Result<FunctionAST, ParseError> {
    expect(tokens, idx, Token::Def)?;
    let proto = parse_prototype(tokens, idx)?;
    let e = parse_expression(tokens, idx)?;
    Ok(FunctionAST { proto, body: e })
}

fn parse_extern(tokens: &[Token], idx: &mut usize) -> Result<PrototypeAST, ParseError> {
    expect(tokens, idx, Token::Extern)?;
    parse_prototype(tokens, idx)
}

fn parse_top_level_expr(tokens: &[Token], idx: &mut usize) -> Result<FunctionAST, ParseError> {
    let e = parse_expression(tokens, idx)?;
    let proto = PrototypeAST {
        name: "__anon_expr".to_string(),
        args: vec![],
    };
    Ok(FunctionAST { proto, body: e })
}

pub fn parse(tokens: &[Token]) -> Result<Vec<FunctionAST>, ParseError> {
    let mut idx = 0;
    let mut fns = Vec::new();
    loop {
        match tokens[idx] {
            Token::Eof => {
                return Ok(fns);
            }
            Token::Semicolon => idx += 1,
            Token::Def => {
                let d = parse_definition(tokens, &mut idx)?;
                fns.push(d);
            }
            Token::Extern => {
                let ext = parse_extern(tokens, &mut idx)?;
                fns.push(FunctionAST {
                    proto: ext,
                    body: ExprAST::Number(0.0),
                });
            }
            _ => {
                let top_level = parse_top_level_expr(tokens, &mut idx)?;
                fns.push(top_level);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::lex;

    #[test]
    fn test_parse_exp() {
        let program: Vec<char> = "x+(1+2)*3-f(x)\n".chars().collect();
        let tokens = lex(&program);
        let mut idx = 0;
        let expr_res = parse_expression(&tokens, &mut idx);
        println!("{:?}", expr_res);
    }
}
